File: 10.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    const dp = [];
    for(let i = 0; i <= s.length; i++) {
        dp.push([]);
        for(let j = 0; j <= p.length; j++) {
            dp[i].push(null);
        }
    }
    const helper = (i, j) => {
        let ans;
        if(dp[i][j] !== null) {
            return dp[i][j];
        }
        if(j === p.length) {
            if(i === s.length) {
                ans = true;
            } else {
                ans = false;
            }
        } else {
            // Pattern p has not run out.
            if(i === s.length) {
                // If s has run out, the only possiblity is
                // p is composed of character *.
                if(j + 1 < p.length && p[j + 1] === '*') {
                    ans = helper(i, j + 2);
                } else {
                    ans = false;
                }
            } else {
                if(s[i] === p[j] || p[j] === '.') {
                    if(j + 1 < p.length && p[j + 1] === '*') {
                        ans = helper(i + 1, j) || helper(i, j + 2);
                    } else {
                        ans = helper(i + 1, j + 1);
                    }
                } else {
                    if(j + 1 < p.length && p[j + 1] === '*') {
                        ans = helper(i, j + 2);
                    } else {
                        ans = false;
                    }
                }
            }
        }
        dp[i][j] = ans;
        return ans;
    };
    return helper(0, 0);
};

File: 10.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        memo = {}
        # Compare s[i:] and p[j:]
        def dp(i, j):
            if (i, j) not in memo:
                if j == len(p):
                    # When pattern has run out but string has not,
                    # the answer should be False.
                    # If both the pattern and the string has run out,
                    # the answer is True.
                    if i == len(s):
                        # s[i:] = ''
                        ans = True
                    else:
                        # s[i:] != ''
                        ans = False
                else:
                    # When pattern has not run out
                    if i < len(s):
                        # First check if pattern starts with .
                        # or the first character of pattern equals 
                        # the first character of the string.
                        if p[j] == '.' or p[j] == s[i]:
                            if j + 1 < len(p) and p[j + 1] == '*':
                                # In this case, we can move one position on the 
                                # string or we can move two position on the pattern.
                                ans = dp(i, j + 2) or dp(i + 1, j)
                            else:
                                # In this case, we can only move both pointer one 
                                # position.
                                ans = dp(i + 1, j + 1)
                        else:
                            # If pattern does not starts with .
                            # and the first character of pattern does not
                            # equals the first character of the string,
                            # the only possible way to be true is 
                            # the pattern starts with character*,
                            # so we can ignore the character*.
                            if j + 1 < len(p) and p[j + 1] == '*':
                                ans = dp(i, j + 2)
                            else:
                                ans = False
                    # When string run out, 
                    # there is still posibility to be True,
                    # which is that last pattern is composed of character*.
                    # So we look at p[j + 1]
                    else:
                        if j + 1 < len(p) and p[j + 1] == '*':
                            ans = dp(i, j + 2)
                        else:
                            ans = False
                memo[(i, j)] = ans
            return memo[(i, j)]
        return dp(0, 0)
                    
                
        